airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/types/GenericParamType.java

```diff
@@ -0,0 +1,28 @@
+package io.airbyte.integrations.types;
+
+/**
+ * This class used as the generic type parameter to define the type of the parameter.
+ * Could be useful fot specific situations when we need to have a generic type parameter
+ *
+ * @see io.airbyte.integrations.destination.redshift.RedshiftDestination
+ */
+public class GenericParamType<T> {
```

Comment:
not sure I understand why GenericParamType is needed, rather than just declaring methods like `onDestinationStartOperations(final JdbcDatabase database, final List<WriteConfig> writeConfigsList)`. I.e. could we just use `T` directly, and not need to wrap it inside this object? We're never actually calling `set`, so they should be equivalent?

(also, I would feel nervous about exposing `set` anyway - that feels like it could make future code pretty complex, since this is basically equivalent to passing a pointer into a method)

Reply:
The `GenericParamType` is used to make SqlOperations more generic and allow client code to choose the second arguments for the next methods:

- onDestinationStartOperations
- onCloseTransactionOperations

This could be extended in the future. BTW we can use Generics by its-self as you suggested, however in such case we need also to change class declaration signature to smth like:

`public class DatabricksSqlOperations<T, S> extends JdbcSqlOperations<T, S>`

instead of

`public class DatabricksSqlOperations extends JdbcSqlOperations<GenericParamType, GenericParamType>  `
(less copy-paste)

and do that for every class extends `JdbcSqlOperations`
This is the basic concept for some Generic "Box" it could be improved and could be more use-cases off-course.

Reply:
I guess I'm still a little unsure why these types need to be parameterized at all - couldn't `onDestinationStartOperations` and `onCloseTransactionOperations` both just accept `Set<String> schemaNames`?

It looks like `JdbcBufferedConsumerFactory`'s call to `onDestinationStartOperations` doesn't declare the generic type at all, so this current implementation is equivalent to just declaring `void onDestinationStartOperations(JdbcDatabase database, Object params)`, which seems pretty problematic.

If we find, in the future, that this method needs to be more flexible in what arguments it accepts, then I think I would actually prefer the raw generic type. It makes it very explicit that they're not used, rather than hiding it behind a raw generic reference (which can be dangerous).

Reply:
@edgao 
Ok, it's clear. I left only 
https://github.com/airbytehq/airbyte/blob/887e29c5367805fa130c2ee24d56f4cb3f4e111f/airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/SqlOperations.java#L134-L134

Moved it to onClose() method in both `JdbcBufferedConsumerFactory` and `CopyConsumerFactory`
I think this will be a reliable solution as it will execute only after:
https://github.com/airbytehq/airbyte/blob/887e29c5367805fa130c2ee24d56f4cb3f4e111f/airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/copy/CopyConsumerFactory.java#L159-L163

And
https://github.com/airbytehq/airbyte/blob/887e29c5367805fa130c2ee24d56f4cb3f4e111f/airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/JdbcBufferedConsumerFactory.java#L160-L163

So, we don't need to update the data column in case of some failures.

Reply:
I think it would be really valuable to add an integration test that validates this migration logic. My understanding is that for standard inserts, the table needs to be updated to use the SUPER type before we start inserting records into it, but I'm not certain.

Reply:
@edgao such test was added here: 
`airbyte-integrations/connectors/destination-redshift/src/test-integration/java/io/airbyte/integrations/destination/redshift/RedshiftInsertDestinationAcceptanceTest.java`
also I have added scenario for the staging
`airbyte-integrations/connectors/destination-redshift/src/test-integration/java/io/airbyte/integrations/destination/redshift/RedshiftS3StagingInsertDestinationAcceptanceTest.java`

Reply:
nice, thanks!